File: Sem4/MPP/Lab/ProiectCSharp/networking/ProjectClientWorker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using model;
using Newtonsoft.Json;
using services;

namespace networking
{
    public class ProjectClientWorker : IProjectObserver
    {
        private IProjectServices server;
        private TcpClient connection;

        private NetworkStream stream;
        private IFormatter formatter;
        private volatile bool connected;

        public ProjectClientWorker(IProjectServices server,TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }
            catch (Exception e) 
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public void ticketBought()
        {
            sendResponse(new BoughtTicketResponse());
        }

        private void sendResponse(Response response)
        {
            Console.WriteLine("Sending response " + response);
            lock (stream)
            {
                formatter.Serialize(stream, response);
                stream.Flush();
            }
        }

        public virtual void run()
        {
            while (connected)
            {
                try
                {
                    object request = formatter.Deserialize(stream);
                    object response = handleRequest((Request)request);
                    if (response != null)
                    {
                        sendResponse((Response)response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error " + e);
            }
        }

        private Response handleRequest(Request request)
        {
            Response response = null;

            if (request is LoginRequest)
            {
                Console.WriteLine("Login request");
                LoginRequest loginRequest = (LoginRequest)request;
                UserDTO userdto = loginRequest.userDTO;
                User user = DTOUtils.fromDTO(userdto);
                try
                {
                    lock (server)
                    {
                        server.login(user, this);
                    }
                    return new OkResponse();
                }
                catch (ProjectException e)
                {
                    connected = false;
                    return new ErrorResponse(e.Message);
                }
            }


            if (request is LogoutRequest)
            {
                Console.WriteLine("Logout request");
                LogoutRequest logoutRequest = (LogoutRequest)request;
                UserDTO userdto = logoutRequest.userDTO;
                User user = DTOUtils.fromDTO(userdto);
                try
                {
                    lock (server)
                    {
                        server.logout(user, this);
                    }
                    connected = false;
                    return new OkResponse();
                }
                catch (ProjectException e)
                {
                    return new ErrorResponse(e.Message);
                }
            }


            if (request is GetTripsRequest)
            {
                Console.WriteLine("GetTrips request");
                GetTripsRequest trips = (GetTripsRequest)request;
                try
                {
                    Excursie[] excursii;
                    lock (server)
                    {
                        excursii = server.getExcursii(this);
                    }
                    ExcursieDTO[] excursiidto = DTOUtils.getDTO(excursii);

                    return new TripsResponse(excursiidto);
                }catch (ProjectException e)
                {
                    return new ErrorResponse(e.Message);
                }
            }


            if (request is GetFilteredTripsRequest)
            {
                Console.WriteLine("GetFilteredTrips request");
                GetFilteredTripsRequest trips = ( GetFilteredTripsRequest)request;
                ExcursieDTO excdto = trips.excursieDTO;
                Excursie exc = DTOUtils.fromDTO(excdto);
                int oramin = trips.oramin;
                int oramax = trips.oramax;
                try
                {
                    Excursie[] excursii;
                    lock (server)
                    {
                        excursii = server.getExcursiiLaLocSiOra(exc, oramin, oramax, this);
                    }
                    ExcursieDTO[] excursiidto = DTOUtils.getDTO(excursii);
                    return new TripsResponse(excursiidto);
                }
                catch (ProjectException e)
                {
                    return new ErrorResponse (e.Message);
                }
            }


            if (request is BuyTicketRequest)
            {
                Console.WriteLine("BuyTicket request");
                BuyTicketRequest buyRequest = ( BuyTicketRequest)request;
                Bilet bilet = DTOUtils.fromDTO(buyRequest.biletDTO);
                try
                {
                    lock (server)
                    {
                        server.rezerva(bilet, this);
                    }
                    ticketBought();
                    return new OkResponse();
                }
                catch(ProjectException e)
                {
                    return new ErrorResponse(e.Message);
                }
            }
            return response;
        }

    }
}

File: Sem4/MPP/Lab/ProiectJava/ProjectNetworking/src/main/java/project/network/jsonprot/ProjectClientWorker.java
package project.network.jsonprot;

import com.google.gson.Gson;
import project.model.Bilet;
import project.model.Excursie;
import project.model.User;
import project.network.dto.BiletDto;
import project.network.dto.DtoUtils;
import project.network.dto.ExcursieDto;
import project.network.dto.UserDto;
import project.services.IObserver;
import project.services.IService;
import project.services.ProjectException;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;

public class ProjectClientWorker implements Runnable, IObserver {
	private IService server;
	private Socket connection;

	private BufferedReader input;
	private PrintWriter output;
	private Gson gsonFormatter;
	private volatile boolean connected;
	public ProjectClientWorker(IService server, Socket connection) {
		this.server = server;
		this.connection = connection;
		gsonFormatter=new Gson();
		try{
			output=new PrintWriter(connection.getOutputStream());
			input=new BufferedReader(new InputStreamReader(connection.getInputStream()));
			connected=true;
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public void run() {
		while(connected){
			try {
				String requestLine=input.readLine();
				Request request=gsonFormatter.fromJson(requestLine, Request.class);
				Response response=handleRequest(request);
				if (response!=null){
					sendResponse(response);
				}
			} catch (IOException e) {
				e.printStackTrace();
			}
			try {
				Thread.sleep(1000);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		try {
			input.close();
			output.close();
			connection.close();
		} catch (IOException e) {
			System.out.println("Error "+e);
		}
	}
	private static Response okResponse=JsonProtocolUtils.createOkResponse();

	private Response handleRequest(Request request){
	Response response = new Response();
		if (request.getType()== RequestType.LOGIN){
			System.out.println("Login request ..."+request.getType());
			UserDto udto=request.getUser();
			User user= DtoUtils.getFromDto(udto);
			try {
				server.login(user, this);
				return okResponse;
			} catch (ProjectException e) {
				connected=false;
				return JsonProtocolUtils.createErrorResponse(e.getMessage());
			}
		}

		if (request.getType()== RequestType.LOGOUT){
			System.out.println("Logout request");

			UserDto udto=request.getUser();
			User user=DtoUtils.getFromDto(udto);
			try {
				server.logout(user, this);
				connected = false;
				return okResponse;
			} catch (ProjectException e) {
				return JsonProtocolUtils.createErrorResponse(e.getMessage());
			}
		}

		if (request.getType()== RequestType.GET_TRIPS){
			System.out.println("GET_TRIPS request");

			try {
				Excursie[] excursii = server.getExcursii(this);
				ExcursieDto[] exdto = DtoUtils.getDto(excursii);
				response.setExcursii(exdto);
				response.setType(ResponseType.TRIPS);
				return response;
			} catch (ProjectException e) {
				return JsonProtocolUtils.createErrorResponse(e.getMessage());
			}
		}

		if (request.getType()== RequestType.GET_FILTERED_TRIPS){
			System.out.println("GET_FILTERED_TRIPS request");

			ExcursieDto excdto = request.getExcursie();
			Excursie exc = DtoUtils.getFromDto(excdto);
			Integer min = request.getOramin();
			Integer max = request.getOramax();
			try {
				Excursie[] excursii = server.getExcursiiLaLocSiOra(exc,min,max,this);
				ExcursieDto[] exdto = DtoUtils.getDto(excursii);
				response.setExcursii(exdto);
				response.setType(ResponseType.FILTERED_TRIPS);
				return response;
			} catch (ProjectException e) {
				return JsonProtocolUtils.createErrorResponse(e.getMessage());
			}
		}

		if (request.getType()== RequestType.BUY_TICKET){
			System.out.println("BUY_TICKET request");

			BiletDto bildto = request.getBilet();
			Bilet bilet = DtoUtils.getFromDto(bildto);
			try {
				server.rezerva(bilet,this);
				return response;
			} catch (ProjectException e) {
				return JsonProtocolUtils.createErrorResponse(e.getMessage());
			}
		}

		return response;
	}

	private void sendResponse(Response response) throws IOException{
		String responseLine=gsonFormatter.toJson(response);
		System.out.println("sending response "+responseLine);
		synchronized (output) {
			output.println(responseLine);
			output.flush();
		}
	}

	public void ticketBought() {
		//Response response = JsonProtocolUtils.create
		Response resp = JsonProtocolUtils.createBoughtTicketResponse();
		System.out.println("Added bilet");
		try {
			sendResponse(resp);
		}catch (IOException e){
			e.printStackTrace();
		}
	}
}
